### DISCUSSION TITLE
Using Tabulator with Objects

### DISCUSSION THREAD

**Discussion by danbolinson-bt** [anvilistas/tabulator#67] [Q&A] [answered]
I have a situation where I want to pass a list of custom objects to Tabulator.

All the relevant data is accessible as though the object were a dictionary, i.e. ```my_obj[field]``` is implemented for all fields.

However, Tabulator cells are empty, and ```cell.getValue()``` returns ```None``` for these fields. 

I can 'force' with a formatter so that data certainly exists: 
```lambda cell: cell.getRow().getData()[cell.getField()],```

Is there a better way to handle custom objects? Or am I doing something wrong here?

Clone link:
https://anvil.works/build#clone:VA3VTMAVOVTEGOI3=W2EKAGTD23TC2J3SRQBMEJUG

**Comment by s-cork** [comment, marked as answer]
Hi @danbolinson-bt , I won't be able to look at the clone till next week - I'm away right now. 

Until then you might want to look at the section in the readme about using models. You can supply a custom getter, which can be arbitrarily complex. Eg. 

```python
def custom_getter(obj, field):
    if field == 'foo':
        return str(obj[field])
    return obj[field]
```

**Reply by s-cork** [reply]
Yeah just had a quick look at the clone. The `use_model` option is exactly what you need. 

The issue is that the Animal class is being converted to an obscure wrapped python object in JavaScript land. In JS the attribute access isn't failing (because JavaScript rarely throws errors) instead it's just getting undefined - hence the None values.

**Reply by danbolinson-bt** [reply]
Okay, sounds good - thanks.
